File: android/app/src/main/java/com/memorize/network/TextService.kt
package com.memorize.network

import com.memorize.database.MemorizeDatabase
import com.memorize.database.entity.*
import java.util.UUID

class TextService(
    private val database: MemorizeDatabase,
    private val deepSeekService: DeepSeekService,
    private val userId: String,
    private val context: android.content.Context? = null
) {
    suspend fun getTextByTitle(title: String, author: String? = null): String? {
        android.util.Log.d("TextService", "getTextByTitle called - title: '$title', author: '$author'")
        return try {
            // Get text from DeepSeek without saving
            val result = deepSeekService.getTextByTitle(title, author)
            android.util.Log.d("TextService", "getTextByTitle result: ${if (result != null) "text length ${result.length}" else "null"}")
            result
        } catch (e: Exception) {
            android.util.Log.e("TextService", "Error in getTextByTitle", e)
            throw e // Re-throw to be handled by SearchViewModel
        }
    }
    
    suspend fun loadAndSaveText(title: String, author: String? = null): String? {
        // Check if text already exists for this user
        val existingText = database.textDao().getTextByTitle(userId, title)
        if (existingText != null) {
            return existingText.id
        }
        
        // Get text from DeepSeek
        val fullText = deepSeekService.getTextByTitle(title, author) ?: return null
        
        // Parse and save
        return saveTextFromFullText(title, fullText)
    }
    
    suspend fun saveTextFromFullText(title: String, fullText: String): String? {
        android.util.Log.d("TextService", "saveTextFromFullText: title='$title', text length=${fullText.length}, userId=$userId")
        
        // Check if text already exists for this user
        val existingText = database.textDao().getTextByTitle(userId, title)
        if (existingText != null) {
            android.util.Log.d("TextService", "saveTextFromFullText: Text already exists, returning existing id: ${existingText.id}")
            return existingText.id
        }
        
        // Save text first
        val textId = UUID.randomUUID().toString()
        val textEntity = TextEntity(
            id = textId,
            userId = userId,
            title = title,
            fullText = fullText
        )
        database.textDao().insertText(textEntity)
        android.util.Log.d("TextService", "saveTextFromFullText: Text entity saved with id: $textId")
        
        // Create simple structure immediately (fast, no API call)
        android.util.Log.d("TextService", "saveTextFromFullText: Creating simple structure...")
        val sentences = fullText.split(Regex("[.!?]+")).filter { it.trim().isNotEmpty() }
        if (sentences.isNotEmpty()) {
            // Get phrases per paragraph from settings (default: 4)
            val phrasesPerParagraph = context?.let {
                it.getSharedPreferences("memorize_prefs", android.content.Context.MODE_PRIVATE)
                    .getInt("phrases_per_paragraph", 4)
            } ?: 4
            
            android.util.Log.d("TextService", "saveTextFromFullText: Using $phrasesPerParagraph phrases per paragraph")
            
            val sectionId = UUID.randomUUID().toString()
            val sectionEntity = SectionEntity(
                id = sectionId,
                textId = textId,
                order = 0
            )
            database.sectionDao().insertSection(sectionEntity)
            
            // Split sentences into paragraphs based on phrasesPerParagraph setting
            sentences.chunked(phrasesPerParagraph).forEachIndexed { paragraphIndex, paragraphSentences ->
                val paragraphId = UUID.randomUUID().toString()
                val paragraphEntity = ParagraphEntity(
                    id = paragraphId,
                    sectionId = sectionId,
                    order = paragraphIndex
                )
                database.paragraphDao().insertParagraph(paragraphEntity)
                
                paragraphSentences.forEachIndexed { phraseIndex, sentence ->
                    val phraseId = UUID.randomUUID().toString()
                    val phraseEntity = PhraseEntity(
                        id = phraseId,
                        paragraphId = paragraphId,
                        order = phraseIndex,
                        text = sentence.trim()
                    )
                    database.phraseDao().insertPhrase(phraseEntity)
                }
            }
            android.util.Log.d("TextService", "saveTextFromFullText: Created structure with ${sentences.size} phrases in ${(sentences.size + phrasesPerParagraph - 1) / phrasesPerParagraph} paragraphs")
        }
        
        android.util.Log.d("TextService", "saveTextFromFullText: Text saved successfully with id: $textId")
        return textId
    }
    
    suspend fun saveTextDirectly(title: String, fullText: String): String? {
        // Use the same method as loadAndSaveText
        return saveTextFromFullText(title, fullText)
    }
}

File: ios/Memorize/Network/TextService.swift
import Foundation

class TextService {
    private let database: MemorizeDatabase
    private let yandexGPTService: YandexGPTService
    private let textRepository: TextRepository
    private let sectionRepository: SectionRepository
    private let paragraphRepository: ParagraphRepository
    private let phraseRepository: PhraseRepository
    
    init(
        database: MemorizeDatabase,
        yandexGPTService: YandexGPTService
    ) {
        self.database = database
        self.yandexGPTService = yandexGPTService
        self.textRepository = TextRepository(database: database)
        self.sectionRepository = SectionRepository(database: database)
        self.paragraphRepository = ParagraphRepository(database: database)
        self.phraseRepository = PhraseRepository(database: database)
    }
    
    func getTextByTitle(title: String, author: String? = nil) async throws -> String? {
        // Get text from Yandex GPT without saving
        return try await yandexGPTService.getTextByTitle(title: title, author: author)
    }
    
    func loadAndSaveText(title: String, author: String? = nil) async throws -> String? {
        // Check if text already exists
        if let existingText = textRepository.getTextById(id: title) {
            return existingText.id
        }
        
        // Get text from Yandex GPT
        guard let fullText = try await yandexGPTService.getTextByTitle(title: title, author: author) else {
            return nil
        }
        
        // Parse text
        guard let parsedStructure = try await yandexGPTService.parseText(fullText) else {
            return nil
        }
        
        // Save to database
        let textId = UUID().uuidString
        let textModel = TextModel(
            id: textId,
            title: title,
            fullText: fullText
        )
        try textRepository.insertText(textModel)
        
        // Save sections, paragraphs, and phrases
        for (sectionIndex, section) in parsedStructure.sections.enumerated() {
            let sectionId = UUID().uuidString
            let sectionModel = SectionModel(
                id: sectionId,
                textId: textId,
                order: sectionIndex
            )
            try sectionRepository.insertSection(sectionModel)
            
            for (paragraphIndex, paragraph) in section.paragraphs.enumerated() {
                let paragraphId = UUID().uuidString
                let paragraphModel = ParagraphModel(
                    id: paragraphId,
                    sectionId: sectionId,
                    order: paragraphIndex
                )
                try paragraphRepository.insertParagraph(paragraphModel)
                
                for (phraseIndex, phraseText) in paragraph.phrases.enumerated() {
                    let phraseId = UUID().uuidString
                    let phraseModel = PhraseModel(
                        id: phraseId,
                        paragraphId: paragraphId,
                        order: phraseIndex,
                        text: phraseText
                    )
                    try phraseRepository.insertPhrase(phraseModel)
                }
            }
        }
        
        return textId
    }
    
    func saveTextDirectly(title: String, fullText: String) async throws -> String? {
        // Check if text already exists
        if let existingText = textRepository.getTextById(id: title) {
            return existingText.id
        }
        
        // Parse text using AI
        guard let parsedStructure = try await yandexGPTService.parseText(fullText) else {
            return nil
        }
        
        // Save to database
        let textId = UUID().uuidString
        let textModel = TextModel(
            id: textId,
            title: title,
            fullText: fullText
        )
        try textRepository.insertText(textModel)
        
        // Save sections, paragraphs, and phrases
        for (sectionIndex, section) in parsedStructure.sections.enumerated() {
            let sectionId = UUID().uuidString
            let sectionModel = SectionModel(
                id: sectionId,
                textId: textId,
                order: sectionIndex
            )
            try sectionRepository.insertSection(sectionModel)
            
            for (paragraphIndex, paragraph) in section.paragraphs.enumerated() {
                let paragraphId = UUID().uuidString
                let paragraphModel = ParagraphModel(
                    id: paragraphId,
                    sectionId: sectionId,
                    order: paragraphIndex
                )
                try paragraphRepository.insertParagraph(paragraphModel)
                
                for (phraseIndex, phraseText) in paragraph.phrases.enumerated() {
                    let phraseId = UUID().uuidString
                    let phraseModel = PhraseModel(
                        id: phraseId,
                        paragraphId: paragraphId,
                        order: phraseIndex,
                        text: phraseText
                    )
                    try phraseRepository.insertPhrase(phraseModel)
                }
            }
        }
        
        return textId
    }
}

